### DISCUSSION TITLE
form.setError with non-string message

### DISCUSSION THREAD

**Discussion by Noitidart** [react-hook-form/react-hook-form#8401] [Q&A] [answered]
Typescript is warning me that form.setError expects a string for the message. However doing the following works fine:

```
form.setError(path, { type: 'react-element', message: <Foo /> )
```

Is this acceptable? I was worried because typescript is warning me, but its working fine.

**Comment by Moshyfawn** [comment, marked as answer]
RHF expects a message to be a `string`. Check out the `setError` `ErrorOption` type:
```ts
export declare type ErrorOption = {
    message?: string
    type?: LiteralUnion<keyof RegisterOptions, string>
    types?: MultipleFieldErrors
}
```

Passing a custom message type, like a `ReactNode` can work fine, as your're the only consumer of the error as a developer, but RHF doesn't support doing this. I doubt it's going to be supported in the future.

That said, there's a helper `ErrorMessage` component form the [error-message](https://github.com/react-hook-form/error-message) package, that expects the message to be a string in order to render it safely. So, just to reiterate: you can use a custom component for your message, but you'll have to mute TS for it, as RHF doesn't support anything but a `string` as an error message, following the Web form API standards.

**Reply by Noitidart** [reply]
Thanks very very much. Oh man ideally if they don't drop support for it, that would be awesome. My actual use case is to give an array but localized parameters like this:

```
setError({
  type: 'localized-error',
  message: ['path.to.key.in.locale.file', { quantity: 3 }]
});
```

I then pass those arguments to the local translation function like `i18n.t(...args)`

**Reply by bluebill1049** [reply]
Yea, this is something bothered me for a long as well, if we do string | React.ReactNode then you would have to do type check string before you get a type error. Allow React.ReactNode or JSX would have made message much more powerful, but this is not a limitation with native JS thanks @Moshyfawn for the detailed answer as usual ❤️

**Reply by Noitidart** [reply]
Thanks very much bluebill for reply!
